tests-legacy/Unit/Classes/ToolsCoreTest.php

```diff
@@ -369,6 +369,48 @@ public function testStrReplaceFirst($search, $replace, $subject, $cur, $expected
         $this->assertEquals($expected, Tools::StrReplaceFirst($search, $replace, $subject, $cur));
     }
 
+    /**
+     * @param string $passwordGenerated
+     * @param string $expectedPassword
+     *
+     * @dataProvider passwodGenProvider
+     */
+    public function testPasswdGen($expectedPassword, $passwordGenerated)
+    {
+        $this->assertRegExp($expectedPassword, $passwordGenerated, 'The password generated ' . $expectedPassword . ' no match with ' . $expectedPassword);
+    }
+
+    public function passwodGenProvider()
+    {
+        return [
+            '/^(\d{8})$/' => Tools::passwdGen(),
+            '/^(\d{13})$/' => Tools::passwdGen(13, Tools::PASSWORDGEN_FLAG_NUMERIC),
+            '/^[A-Z]{9}$/' => Tools::passwdGen(9, Tools::PASSWORDGEN_FLAG_NO_NUMERIC),
+            '/^(\w){14}$/' => Tools::passwdGen(14, Tools::PASSWORDGEN_FLAG_RANDOM),
+            '/^(\w){10}$/' => Tools::passwdGen(10, Tools::PASSWORDGEN_FLAG_ALPHANUMERIC),
+            '/^(false|0)$/' => Tools::passwdGen(1),
```

Comment:
You duplicate the index `'/^(false|0)$/'`, only the last statement is executed.

I'm also afraid that you must use `false` or `0` a regex with false inside is also interpreted as string not only boolean, which mean if you have a password randomly generated the word `false` you're rekt :sweat_smile: 

Regarding the code, you should test for `false` or `0` per tests, not both together, the method can't return `false` and `0`, the if statement is not the same.

Reply:
@PierreRambaud  But `false` is a weak password for `openssl_random_pseudo_bytes`, so it will return `false` (bool) and not `string`.